### PR TITLE
Adds a worker script for running quality control jobs.

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

File: jenkins/worker.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
+import argparse
```
Comment (SirVer) [inline jenkins/worker.py]: Run `pyformat` and `gpylint` and fix complains.

File: jenkins/worker.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+                            r'(?P<minutes>\d{1,2}):(?P<seconds>\d{2}\.\d{2})')
+MAX_RES_SET_SIZE_PATTERN = Pattern(r'^\s*Maximum resident set size \(kbytes\): (?P<max_set_size>\d+)')
+
+
+# Pattern matcher for extracting the HEAD commit SHA-1 hash.
+GIT_SHA1_PATTERN = Pattern(r'^(?P<sha1>[0-9a-f]{40})\s+HEAD')
+
+
+# Returns the SHA-1 hash of the commit tagged HEAD.
+def get_tot_git_sha1():
```
Comment (SirVer) [inline jenkins/worker.py]: google style guide: CapitalNames. Also always below.
https://google.github.io/styleguide/pyguide.html

File: jenkins/worker.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    output = subprocess.check_output(['git', 'ls-remote', 'https://github.com/googlecartographer/cartographer.git'])
```
Comment (SirVer) [inline jenkins/worker.py]: 80 columns. Also a few times below.

File: jenkins/worker.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    dict = GIT_SHA1_PATTERN.extract(output)
+    return dict['sha1']
+
+
+# Returns a dictionary of stats.
```
Comment (SirVer) [inline jenkins/worker.py]: docstring instead of comment. Also always below.

File: jenkins/worker.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+    dict = USER_TIME_PATTERN.extract(input)
+    result['user_time_secs'] = float(dict['user_time'])
+
+    dict = SYSTEM_TIME_PATTERN.extract(input)
+    result['system_time_secs'] = float(dict['system_time'])
+
+    dict = WALL_TIME_PATTERN.extract(input)
+    wall_time_secs = 0.0
+    if (dict['hours'] != None):
```
Comment (SirVer) [inline jenkins/worker.py]: `wall_time_secs` = float(dict.get('hours', 0.)) * 3600 + float(dict['minutes']) * 60 + float(dict['seconds'])

File: jenkins/worker.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        wall_time_secs += int(dict['hours']) * 3600
+    wall_time_secs += int(dict['minutes']) * 60
```
Comment (SirVer) [inline jenkins/worker.py]: make it float right away since you are converting `wall_time_secs` to float anyways.

File: jenkins/worker.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+
+# Constructs a round-robin job selector.
+def create_job_selector(worker_id, num_workers):
+    return lambda job_id: job_id % num_workers == worker_id
+
+
+# Runs the provided command both printing its stdout output to stdout and returning the output as a string.
+def run_cmd(cmd):
+    p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```
Comment (SirVer) [inline jenkins/worker.py]: p.communicate is not good enough? I guess you want this to stream.

File: jenkins/worker.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+class Job:
+    def __init__(self, datastore_client, job_id):
+        self.id = job_id
+        entity = retrieve_entity(datastore_client, "Job", job_id)
+        self.launch_file = entity['launch_file']
+        self.assets_writer_launch_file = entity['assets_writer_launch_file']
+        self.assets_writer_config_file = entity['assets_writer_config_file']
+        self.rosbag = entity['rosbag']
+
+    def __str__(self):
```
Comment (SirVer) [inline jenkins/worker.py]: should this not be `__repr__`?  Also, add the class name into this

File: jenkins/worker.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+        info['rosbag'] = rosbag_filename
+        return info
+
+
+# Represents a single worker that executes a sequence of Jobs.
+class Worker:
+    def __init__(self, datastore_client, pipeline_id, run_id):
+        entity = retrieve_entity(datastore_client, "PipelineConfig", pipeline_id)
+        self.pipeline_id = pipeline_id
+        self.jobs = map(functools.partial(Job, datastore_client), entity['jobs'])
```
Comment (SirVer) [inline jenkins/worker.py]: Use a list comprehension for better readability.

File: jenkins/worker.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+    parser.add_argument('--num_workers', type=int)
+    parser.add_argument('--pipeline_id', type=str)
+
+    args = parser.parse_args()
+
+    datastore_client = datastore.Client()
+    bigquery_client = bigquery.Client()
+    dataset = bigquery_client.dataset('Cartographer')
+    table = dataset.table('metrics')
+
```
Comment (SirVer) [inline jenkins/worker.py]: cut down on multiple empty lines.

File: jenkins/worker.py [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+            if job_selector(idx):
+                output = job.run(self.ros_distro, self.run_id)
+                outputs[job.id] = output
+            else:
+                print 'job {}: skip'.format(job.id)
+        return outputs
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
```
Comment (SirVer) [inline jenkins/worker.py]: Pull out a function Main() and just call it here.

File: jenkins/worker.py [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument('--worker_id', type=int)
+    parser.add_argument('--num_workers', type=int)
+    parser.add_argument('--pipeline_id', type=str)
+
+    args = parser.parse_args()
```
Comment (SirVer) [inline jenkins/worker.py]: Also pull out Argument Parsing into a separate function.

File: jenkins/worker.py (hunk 2 above)
Comment (SirVer) [inline jenkins/worker.py]: Prefer HEAD to `tot`

File: jenkins/worker.py [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+
+
+    job_selector = create_job_selector(int(args.worker_id),
+                                        int(args.num_workers))
+
+    tot_sha1 = get_tot_git_sha1()
+    now = datetime.datetime.now()
+    run_id = '{}-{}-{}_{}'.format(now.year, now.month, now.day, tot_sha1)
+
+    worker = Worker(datastore_client,
```
Comment (SirVer) [inline jenkins/worker.py]: fits in 80 cols

File: jenkins/worker.py [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+                    args.pipeline_id,
+                    run_id)
+
+
+    # Publish stats to BigQuery.
```
Comment (SirVer) [inline jenkins/worker.py]: Pull out into function.

File: jenkins/worker.py [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+    stats_dict = worker.run_jobs(job_selector)
+    rows = []
+    for job_id, info in stats_dict.iteritems():
+        print info
+        data_string = '''[
```
Comment (SirVer) [inline jenkins/worker.py]: Add a comment that order matters here.

File: jenkins/worker.py (hunk 1 above)
Comment (cschuet) [reply jenkins/worker.py]: Done.

File: jenkins/worker.py (hunk 2 above)
Comment (cschuet) [reply jenkins/worker.py]: This changed. function_name() is ok.

Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 3 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 4 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 5 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 6 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 7 above)
Comment (cschuet) [reply jenkins/worker.py]: right, I want lines to appear in my Jenkins UI as soon as they are printed.

File: jenkins/worker.py (hunk 8 above)
Comment (cschuet) [reply jenkins/worker.py]: Done

File: jenkins/worker.py (hunk 9 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 11 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 12 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 10 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 13 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py (hunk 14 above)
Comment (cschuet) [reply jenkins/worker.py]: done

File: jenkins/worker.py [hunk 16]
```diff
@@ -1,204 +1,196 @@
...
-    # Returns a dictionary of named capture groups to extracted output. Returns an empty dict of no match was found.
-    def extract(self, input):
-        match = self.regex.search(input)
-        if match == None:
-            return {}
-        return match.groupdict()
+  def __init__(self, pattern):
+    self.regex = re.compile(pattern, re.MULTILINE)
 
+  # Returns a dictionary of named capture groups to extracted output. Returns an
```
Comment (SirVer) [inline jenkins/worker.py]: docstring

File: jenkins/worker.py [hunk 17]
```diff
@@ -1,204 +1,196 @@
...
+    # Creates assets.
+    run_cmd('/bin/bash -c \"source /opt/ros/{}/setup.bash && source '
+            '/opt/cartographer_ros/setup.bash && /usr/bin/time -v roslaunch '
+            'cartographer_ros {} bag_filenames:={}/{} '
+            'pose_graph_filename:={}/{}.pbstream config_file:={}\"'.format(
+                ros_distro, self.assets_writer_launch_file, scratch_dir,
+                rosbag_filename, scratch_dir, rosbag_filename,
+                self.assets_writer_config_file))
+
+    # Copy assets to bucket.
```
Comment (SirVer) [inline jenkins/worker.py]: Consistency: Copies

File: jenkins/worker.py [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+  """Constructs a round-robin job selector."""
+  return lambda job_id: job_id % num_workers == worker_id
+
+
+def run_cmd(cmd):
+  """Runs command both printing its stdout output and returning it as string."""
+  p = subprocess.Popen(
+      cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+  output = ''
+  while p.poll() is None:
```
Comment (SirVer) [inline jenkins/worker.py]: for line in p.stdout: should do the same thing or not? Poll should not be necessary. 

Also, collect lines into a list and '\n'.join() at the end.

Comment (cschuet) [reply jenkins/worker.py]: as discussed changed.

File: jenkins/worker.py (hunk 17 above)
Comment (cschuet) [reply jenkins/worker.py]: Done

File: jenkins/worker.py (hunk 18 above)
Comment (cschuet) [reply jenkins/worker.py]: Done

**Pull request merged**